Fix binary op name inference to happen before shape checks

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25563 Fix binary op name inference to happen before shape checks**

Before, for binary ops, name inference occurred after shape checks. This
defeats the purposes for names because the names are supposed to tell
the user that i.e. their tensors are misaligned or that they are adding
incompatible tensors.

This PR changes TensorIterator so that names are computed before shape checks and
propagated after the binary ops are finished. In order to support this,
this PR makes the following changes:
- adds a `names_` field to TensorIterator, similar to `shape_`. This is
necessary to hold the output names, that are computed in
`compute_names`, until they are used in `propagate_names_to_outputs()`.

Differential Revision: [D17158869](https://our.internmc.facebook.com/intern/diff/D17158869)

zou3519 committed:
Fix binary op name inference to happen before shape checks

Before, for binary ops, name inference occurred after shape checks. This
defeats the purposes for names because the names are supposed to tell
the user that i.e. their tensors are misaligned or that they are adding
incompatible tensors.

This PR changes TensorIterator so that names are computed before shape checks and
propagated after the binary ops are finished. In order to support this,
this PR makes the following changes:
- adds a `names_` field to TensorIterator, similar to `shape_`. This is
necessary to hold the output names, that are computed in
`compute_names`, until they are used in `propagate_names_to_outputs()`.

nairbv commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -214,40 +214,45 @@ void TensorIterator::allocate_outputs() {
+  }
 
   // propagate names
   for (int i = 0; i < num_outputs_; i++) {
```
nit: not really related to your change but I might prefer `i < noutputs()` and `outputs(i)` here.

zou3519 replied:
Glancing through the TensorIterator code, I see that `num_outputs_` is used a lot instead of `noutputs()`. I'm not sure what the concerns are, maybe `noutputs()` is more of the outward-facing API.

`outputs(i)` performs an unnecessary `AT_ASSERT`. We don't compile those away and until we do, putting it in a loop like this probably won't let us squeeze out the last bit of performance that we're trying to get.

nairbv replied:
I just prefer it because `operands_[i]` may not be clear to some readers since it relies on knowing that outputs are stored at the beginning of the operands array. I'm not sure what the perf overhead is and we do use `operands_[i]` elsewhere here, so not a review-blocker, just a preference of mine.

nairbv commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -214,40 +214,45 @@ void TensorIterator::allocate_outputs() {
-void TensorIterator::propagate_names_to_outputs() {
-  bool should_perform_name_inference = std::any_of(
+void TensorIterator::compute_names() {
+  bool should_infer_names = std::any_of(
```
I see you renamed this variable, and then removed the usage of the prior name, ... is this used somewhere?

zou3519 replied:
Oh, that was an accident. It should be used right after. Thanks for catching

zou3519 committed:
Update on "Fix binary op name inference to happen before shape checks"


Before, for binary ops, name inference occurred after shape checks. This
defeats the purposes for names because the names are supposed to tell
the user that i.e. their tensors are misaligned or that they are adding
incompatible tensors.

This PR changes TensorIterator so that names are computed before shape checks and
propagated after the binary ops are finished. In order to support this,
this PR makes the following changes:
- adds a `names_` field to TensorIterator, similar to `shape_`. This is
necessary to hold the output names, that are computed in
`compute_names`, until they are used in `propagate_names_to_outputs()`.

Differential Revision: [D17158869](https://our.internmc.facebook.com/intern/diff/D17158869)
